File: ios/TwitterFlux/Sources/Views/ComposeView.swift
// ComposeView — tweet compose page.

import SwiftUI

struct ComposeView: View {
    @EnvironmentObject var store: FluxStore
    @Environment(\.dismiss) var dismiss
    @State private var content = ""

    private var compose: ComposeState? { store.get("compose/state") }
    private var charCount: Int { content.count }
    private var isOverLimit: Bool { charCount > 280 }

    var body: some View {
        NavigationStack {
            VStack(spacing: 0) {
                TextEditor(text: $content)
                    .padding()
                    .onChange(of: content) { _, newValue in
                        store.emit("compose/update-field", json: [
                            "field": "content",
                            "value": newValue,
                        ])
                    }

                Divider()

                HStack {
                    Text(store.t("format/char_count?current=\(charCount)&max=280"))
                        .font(.caption)
                        .foregroundColor(isOverLimit ? .red : .secondary)

                    Spacer()

                    if let error = compose?.error {
                        Text(error)
                            .font(.caption)
                            .foregroundColor(.red)
                    }
                }
                .padding()
            }
            .navigationTitle(store.t("ui/compose/title"))
            #if os(iOS)
            .navigationBarTitleDisplayMode(.inline)
            #endif
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button(store.t("ui/compose/cancel")) { dismiss() }
                }
                ToolbarItem(placement: .confirmationAction) {
                    Button(store.t("ui/compose/post")) { postTweet() }
                        .bold()
                        .disabled(content.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
                                  || isOverLimit
                                  || compose?.busy == true)
                }
            }
        }
    }

    private func postTweet() {
        store.emit("tweet/create", json: ["content": content])
        dismiss()
    }
}

File: e2e/flux/flutter/lib/views/compose_view.dart
/// ComposeView — tweet compose page (mirrors Swift ComposeView).
library;

import 'package:flutter/cupertino.dart';

import '../models/models.dart';
import '../store/flux_store.dart';

class ComposeView extends StatefulWidget {
  const ComposeView({super.key});

  @override
  State<ComposeView> createState() => _ComposeViewState();
}

class _ComposeViewState extends State<ComposeView> {
  final _contentController = TextEditingController();

  int get _charCount => _contentController.text.length;
  bool get _isOverLimit => _charCount > 280;

  @override
  void dispose() {
    _contentController.dispose();
    super.dispose();
  }

  void _post() {
    final store = FluxStoreScope.of(context);
    store.emit('tweet/create', {'content': _contentController.text});
    Navigator.of(context).pop();
  }

  @override
  Widget build(BuildContext context) {
    final store = FluxStoreScope.of(context);
    final compose = store.get<ComposeState>('compose/state');

    final canPost =
        _contentController.text.trim().isNotEmpty &&
        !_isOverLimit &&
        compose?.busy != true;

    return CupertinoPageScaffold(
      navigationBar: CupertinoNavigationBar(
        middle: Text(store.t('ui/compose/title')),
        leading: CupertinoButton(
          padding: EdgeInsets.zero,
          child: Text(store.t('ui/compose/cancel')),
          onPressed: () => Navigator.of(context).pop(),
        ),
        trailing: CupertinoButton(
          padding: EdgeInsets.zero,
          onPressed: canPost ? _post : null,
          child: Text(
            store.t('ui/compose/post'),
            style: const TextStyle(fontWeight: FontWeight.bold),
          ),
        ),
      ),
      child: SafeArea(
        child: Column(
          children: [
            // Text editor area
            Expanded(
              child: Padding(
                padding: const EdgeInsets.all(16),
                child: CupertinoTextField(
                  controller: _contentController,
                  maxLines: null,
                  expands: true,
                  textAlignVertical: TextAlignVertical.top,
                  decoration: const BoxDecoration(),
                  onChanged: (_) => setState(() {}),
                ),
              ),
            ),

            // Divider
            Container(
              height: 1,
              color: CupertinoColors.separator.resolveFrom(context),
            ),

            // Bottom bar: char count + error
            Padding(
              padding: const EdgeInsets.all(16),
              child: Row(
                children: [
                  Text(
                    store.t('format/char_count?current=$_charCount&max=280'),
                    style: TextStyle(
                      fontSize: 12,
                      color: _isOverLimit
                          ? CupertinoColors.destructiveRed
                          : CupertinoColors.secondaryLabel,
                    ),
                  ),
                  const Spacer(),
                  if (compose?.error != null)
                    Text(
                      compose!.error!,
                      style: const TextStyle(
                        fontSize: 12,
                        color: CupertinoColors.destructiveRed,
                      ),
                    ),
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }
}
